File: src/content/docs/topics/cs/coding-problems/graphs/269-alien-dictionary.py
def alien_order(words):
    pass  # TODO: implement

def _run_tests():
    result = alien_order(['wrt', 'wrf', 'er', 'ett', 'rftt'])
    assert result == 'wertf', f'got {result!r}'
    result = alien_order(['z', 'x'])
    assert result == 'zx', f'got {result!r}'
    assert alien_order(['z', 'x', 'z']) == ''
    assert alien_order(['abc', 'ab']) == ''
    result = alien_order(['abc'])
    assert set(result) == set('abc'), f'got {result!r}'
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    big_words = [chr(ord('a') + i % 26) * 3 + chr(ord('a') + (i + 1) % 26) for i in range(200)]
    alien_order(big_words)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf alien-dictionary 200 words: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/graphs/269-alien-dictionary.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func alienOrder(_ words: [String]) -> String {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:2e7ed925801d40c6c21bcfa7f7e854ace7dbcd3d55ba691ed723efc89758e7a5
    expectEqual(Solution().alienOrder(["wrt", "wrf", "er", "ett", "rftt"]), "wertf", "canonical-order")
    expectEqual(Solution().alienOrder(["z", "x", "z"]), "", "cycle-has-no-order")
    expectEqual(Solution().alienOrder(["a"]), "a", "single-word")
    // EXCLUDED_VECTOR empty-dictionary: [[]] | The problem contract requires at least one word.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
